File: code/jvm/im-services/src/main/kotlin/im/services/invitations/InvitationService.kt
package im.services.invitations

import im.domain.Either
import im.domain.channel.ChannelRole
import im.domain.invitations.ChannelInvitation
import im.domain.invitations.ChannelInvitationStatus
import im.domain.invitations.ImInvitation
import im.domain.user.User
import im.domain.wrappers.identifier.Identifier
import im.repository.pagination.Pagination
import im.repository.pagination.PaginationRequest
import im.repository.pagination.SortRequest
import im.services.auth.AuthError
import java.time.LocalDateTime

interface InvitationService {
    /**
     * Creates an invitation for a user to join a channel
     *
     * - The inviter must be the owner of the channel to create an invitation
     * - The invitee must not already be a member of the channel
     * - The invitee must not already have a pending invitation to the channel
     * - The expiration date must be at least 15 minutes in the future
     * - The expiration date must be at most 30 days in the future
     *
     *
     * @param channelId the ID of the channel to invite the user to
     * @param inviteeId the ID of the user to invite
     * @param expirationDate the date when the invitation will expire
     * @param role the role the user will have in the channel
     * @param inviter the user who is inviting the invitee
     *
     * @return the created invitation or an error if the invitation could not be created
     */
    fun createChannelInvitation(
        channelId: Identifier,
        inviteeId: Identifier,
        expirationDate: LocalDateTime,
        role: ChannelRole,
        inviter: User,
    ): Either<InvitationError, ChannelInvitation>

    /**
     * Creates an Instant Messaging invitation.
     *
     *
     * @param expiration the expiration date of the invitation
     * @return the invitation if it is created, or an [AuthError] otherwise
     */
    fun createImInvitation(expiration: LocalDateTime?): Either<InvitationError, ImInvitation>

    /**
     * Gets an invitation's details
     *
     * - The user must be the inviter or the invitee to view the invitation
     *
     * @param channelId the ID of the channel to invite the user to
     * @param inviteId the ID of the invitation to get
     * @param user the user requesting the invitation
     *
     * @return the invitation or an error if the user does not have permission to view the invitation
     */
    fun getInvitation(
        channelId: Identifier,
        inviteId: Identifier,
        user: User,
    ): Either<InvitationError, ChannelInvitation>

    /**
     * Gets invitations for a channel
     *
     * - The user must be the owner of the channel to view the invitations for the channel
     *
     * @param channelId the ID of the channel to get the invitations for
     * @param user the user requesting the invitations
     * @param sortRequest the sort request
     * @param paginationRequest the pagination request
     *
     * @return a list of invitations for the channel or an error if the user does not have permission to view the
     * invitations
     */
    fun getChannelInvitations(
        channelId: Identifier,
        user: User,
        sortRequest: SortRequest,
        paginationRequest: PaginationRequest,
    ): Either<InvitationError, Pagination<ChannelInvitation>>

    /**
     * Updates an invitation for a user to join a channel
     *
     * - The user must be the inviter to update the invitation
     * - The new expiration date must be at least 15 minutes in the future
     * - The new expiration date must be at most 30 days in the future
     *
     * @param channelId the ID of the channel to invite the user to
     * @param inviteId the ID of the invitation to update
     * @param role the role the user will have in the channel
     * @param expirationDate the date when the invitation will expire
     * @param user the user updating the invitation
     *
     * @return an error if the invitation could not be updated
     */
    fun updateInvitation(
        channelId: Identifier,
        inviteId: Identifier,
        role: ChannelRole,
        expirationDate: LocalDateTime,
        user: User,
    ): Either<InvitationError, Unit>

    /**
     * Deletes an invitation for a user to join a channel
     *
     * - The user must be the inviter to delete the invitation
     *
     * @param channelId the ID of the channel to invite the user to
     * @param inviteId the ID of the invitation to delete
     * @param user the user deleting the invitation
     *
     * @return an error if the invitation could not be deleted
     */
    fun deleteInvitation(
        channelId: Identifier,
        inviteId: Identifier,
        user: User,
    ): Either<InvitationError, Unit>

    /**
     * Gets the invitations a user has received
     *
     * - A user can only view the invitations they have received
     *
     * @param userId the ID of the user to get the invitations for
     * @param user the user requesting the invitations
     * @param sortRequest the sort request
     * @param paginationRequest the pagination request
     *
     * @return a list of invitations for the user to accept or an error if the user does not have permission to view the
     * invitations
     */
    fun getUserInvitations(
        userId: Identifier,
        user: User,
        sortRequest: SortRequest,
        paginationRequest: PaginationRequest,
    ): Either<InvitationError, Pagination<ChannelInvitation>>

    /**
     * Accepts or rejects an invitation for a user to join a channel
     *
     * - The user must be the invitee to accept or reject the invitation
     *
     * @param userId the ID of the user to accept or reject the invitation for
     * @param invitationIdentifier the ID of the invitation to accept or reject
     * @param status whether to accept or reject the invitation
     * @param user the user accepting or rejecting the invitation
     *
     * @return an error if the invitation could not be accepted or rejected
     */
    fun acceptOrRejectInvitation(
        userId: Identifier,
        invitationIdentifier: Identifier,
        status: ChannelInvitationStatus,
        user: User,
    ): Either<InvitationError, Unit>
}

File: code/js/src/Services/invitations/InvitationService.ts
import {
    ChannelInvitation,
    channelInvitationFromCreation,
    channelInvitationFromDto,
} from '../../Domain/invitations/ChannelInvitation';
import { Channel } from '../../Domain/channel/Channel';
import { User } from '../../Domain/user/User';
import { ChannelRole } from '../../Domain/channel/ChannelRole';
import { ApiResult, handle } from '../media/Problem';
import { BaseHTTPService } from '../BaseHTTPService';
import { ChannelInvitationCreationInputModel } from '../../Dto/input/ChannelInvitationCreationInputModel';
import { ChannelInvitationCreationOutputModel } from '../../Dto/output/invitations/ChannelInvitationCreationOutputModel';
import { Uri } from '../Uri';
import { Identifier } from '../../Domain/wrappers/identifier/Identifier';
import { ChannelInvitationOutputModel } from '../../Dto/output/invitations/ChannelInvitationOutputModel';
import { PaginationRequest } from '../../Domain/pagination/PaginationRequest';
import { SortRequest } from '../../Domain/pagination/SortRequest';
import { ChannelInvitationsPaginatedOutputModel } from '../../Dto/output/invitations/ChannelInvitationsPaginatedOutputModel';
import { buildQuery } from '../Utils';
import { Pagination } from '../../Domain/pagination/Pagination';
import { ChannelInvitationUpdateInputModel } from '../../Dto/input/ChannelInvitationUpdateInputModel';
import { InvitationAcceptInputModel } from '../../Dto/input/InvitationAcceptInputModel';
import { ChannelInvitationStatus } from '../../Domain/invitations/ChannelInvitationStatus';

export namespace InvitationService {
    import post = BaseHTTPService.post;
    import get = BaseHTTPService.get;
    import patch = BaseHTTPService.patch;
    import deleteRequest = BaseHTTPService.deleteRequest;

    import CHANNEL_INVITATIONS_ROUTE = Uri.CHANNEL_INVITATIONS_ROUTE;
    import CHANNEL_ID_PARAM = Uri.CHANNEL_ID_PARAM;
    import INVITATION_ID_PARAM = Uri.INVITATION_ID_PARAM;
    import CHANNEL_INVITATION_ROUTE = Uri.CHANNEL_INVITATION_ROUTE;
    import USER_INVITATIONS_ROUTE = Uri.USER_INVITATIONS_ROUTE;
    import USER_ID_PARAM = Uri.USER_ID_PARAM;
    import USER_INVITATION_ROUTE = Uri.USER_INVITATION_ROUTE;

    /**
     * Creates a channel invitation.
     *
     * @param channel - The channel to create the invitation for.
     * @param invitee - The user to invite.
     * @param expiresAt - The date the invitation expires.
     * @param role - The role of the user in the channel.
     * @param abortSignal - The signal to abort the request.
     *
     * @returns The channel invitation that was created.
     */
    export async function createChannelInvitation(
        channel: Channel,
        invitee: User,
        expiresAt: Date,
        role: ChannelRole,
        abortSignal?: AbortSignal,
    ): ApiResult<ChannelInvitation> {
        return await handle(
            post<ChannelInvitationCreationInputModel, ChannelInvitationCreationOutputModel>({
                uri: CHANNEL_INVITATIONS_ROUTE.replace(CHANNEL_ID_PARAM, channel.id.value.toString()),
                requestBody: {
                    invitee: invitee.id.value,
                    expiresAt: expiresAt,
                    role: role,
                },
                abortSignal: abortSignal,
            }),
            (outputModel) => channelInvitationFromCreation(outputModel, channel, invitee, expiresAt, role),
        );
    }

    /**
     * Gets a channel invitation by its identifier.
     *
     * @param channel - The channel the invitation belongs to.
     * @param inviteId - The identifier of the invitation.
     * @param abortSignal - The signal to abort the request.
     *
     * @returns The channel invitation.
     */
    export async function getInvitation(
        channel: Channel,
        inviteId: Identifier,
        abortSignal?: AbortSignal,
    ): ApiResult<ChannelInvitation> {
        return await handle(
            get<ChannelInvitationOutputModel>({
                uri: CHANNEL_INVITATION_ROUTE.replace(CHANNEL_ID_PARAM, channel.id.value.toString()).replace(
                    INVITATION_ID_PARAM,
                    inviteId.value.toString(),
                ),
                abortSignal: abortSignal,
            }),
            channelInvitationFromDto,
        );
    }

    /**
     * Gets a list of channel invitations.
     *
     * @param channel - The channel the invitations belong to.
     * @param pagination - The pagination request.
     * @param sort - The sort request.
     * @param after - The identifier of the last item in the previous page.
     * @param abortSignal - The signal to abort the request.
     *
     * @returns The list of channel invitations.
     */
    export async function getChannelInvitations(
        channel: Channel,
        pagination?: PaginationRequest,
        sort?: SortRequest,
        after?: Identifier,
        abortSignal?: AbortSignal,
    ): ApiResult<Pagination<ChannelInvitation>> {
        return await handle(
            get<ChannelInvitationsPaginatedOutputModel>({
                uri: buildQuery(CHANNEL_INVITATIONS_ROUTE, null, pagination, sort, false, after).replace(
                    CHANNEL_ID_PARAM,
                    channel.id.value.toString(),
                ),
                abortSignal: abortSignal,
            }),
            (outputModel) => ({
                items: outputModel.invitations.map(channelInvitationFromDto),
                info: outputModel.pagination,
            }),
        );
    }

    /**
     * Updates a channel invitation.
     *
     * @param invitation - The invitation to update.
     * @param role - The new role of the user in the channel.
     * @param expiresAt - The new expiration date of the invitation.
     * @param abortSignal - The signal to abort the request.
     *
     */
    export async function updateInvitation(
        invitation: ChannelInvitation,
        role: ChannelRole,
        expiresAt: Date,
        abortSignal?: AbortSignal,
    ): ApiResult<void> {
        return patch<ChannelInvitationUpdateInputModel>({
            uri: CHANNEL_INVITATION_ROUTE.replace(CHANNEL_ID_PARAM, invitation.channel.id.value.toString()).replace(
                INVITATION_ID_PARAM,
                invitation.id.value.toString(),
            ),
            requestBody: {
                role: role,
                expiresAt: expiresAt,
            },
            abortSignal: abortSignal,
        });
    }

    /**
     * Deletes a channel invitation.
     *
     * @param invitation - The invitation to delete.
     * @param abortSignal - The signal to abort the request.
     */
    export async function deleteInvitation(invitation: ChannelInvitation, abortSignal?: AbortSignal): ApiResult<void> {
        return await deleteRequest({
            uri: CHANNEL_INVITATION_ROUTE.replace(CHANNEL_ID_PARAM, invitation.channel.id.value.toString()).replace(
                INVITATION_ID_PARAM,
                invitation.id.value.toString(),
            ),
            abortSignal: abortSignal,
        });
    }

    /**
     * Gets a list of invitations for a user.
     *
     * @param user - The user to get the invitations for.
     * @param pagination - The pagination request.
     * @param sort - The sort request.
     * @param after - The identifier of the last item in the previous page.
     * @param abortSignal - The signal to abort the request.
     *
     * @returns The list of invitations.
     */
    export async function getUserInvitations(
        user: User,
        pagination?: PaginationRequest,
        sort?: SortRequest,
        after?: Identifier,
        abortSignal?: AbortSignal,
    ): ApiResult<Pagination<ChannelInvitation>> {
        return await handle(
            get<ChannelInvitationsPaginatedOutputModel>({
                uri: buildQuery(USER_INVITATIONS_ROUTE, null, pagination, sort, false, after)
                    .replace(CHANNEL_ID_PARAM, user.id.value.toString())
                    .replace(USER_ID_PARAM, user.id.value.toString()),
                abortSignal: abortSignal,
            }),
            (outputModel) => ({
                items: outputModel.invitations.map(channelInvitationFromDto),
                info: outputModel.pagination,
            }),
        );
    }

    /**
     * Accepts a channel invitation.
     *
     * @param invitation - The invitation to accept.
     * @param abortSignal - The signal to abort the request.
     *
     */
    export async function acceptInvitation(invitation: ChannelInvitation, abortSignal?: AbortSignal): ApiResult<void> {
        return await patch<InvitationAcceptInputModel>({
            uri: USER_INVITATION_ROUTE.replace(CHANNEL_ID_PARAM, invitation.channel.id.value.toString())
                .replace(INVITATION_ID_PARAM, invitation.id.value.toString())
                .replace(USER_ID_PARAM, invitation.invitee.id.value.toString()),
            requestBody: {
                status: ChannelInvitationStatus.ACCEPTED,
            },
            abortSignal: abortSignal,
        });
    }

    /**
     * Declines a channel invitation.
     *
     * @param invitation - The invitation to decline.
     * @param abortSignal - The signal to abort the request.
     */
    export async function declineInvitation(invitation: ChannelInvitation, abortSignal?: AbortSignal): ApiResult<void> {
        return await patch<InvitationAcceptInputModel>({
            uri: USER_INVITATION_ROUTE.replace(CHANNEL_ID_PARAM, invitation.channel.id.value.toString())
                .replace(INVITATION_ID_PARAM, invitation.id.value.toString())
                .replace(USER_ID_PARAM, invitation.invitee.id.value.toString()),
            requestBody: {
                status: ChannelInvitationStatus.REJECTED,
            },
            abortSignal: abortSignal,
        });
    }
}
